repo_lib/__init__.py

```diff
@@ -0,0 +1,121 @@
+            raise RepoError('component does not exist',
+                            join(self.path, component_dir))
+
+        output_dir = join(self.path, 'dists', self.release,
+                          component, f'binary-{arch}')
+
+        if not exists(output_dir):
+            os.makedirs(output_dir)
+
+        output = self._archive_cmd('packages', component_dir, arch=arch)
+        with open(join(output_dir, 'Packages'), "w") as fob:
+            if output:
+                output += '\n'
+            fob.write(output)
+
+        with open(join(output_dir, 'Packages'), 'rb') as fob:
+            with gzip.open(join(output_dir, 'Packages.gz'), 'wb') as zob:
+                zob.writelines(fob)
+            with bz2.open(join(output_dir, 'Packages.bz2'), 'wb') as zob:
+                zob.writelines(fob)
+
+        with open(join(output_dir, 'Release'), "w") as fob:
+            fob.writelines(
+                [f"Origin: {self.origin}\n",
+                 f"Label: {self.origin}\n",
+                 f"Archive: {self.release}\n",
+                 f"Version: {self.version}\n",
+                 f"Component: {component}\n",
+                 f"Architecture: {arch}\n"])
+
+    def generate_release(self, gpgkey: Optional[str] = None):
+        def get_archs() -> Set[str]:
+            archs = set()
+            dist_path = join(self.path, 'dists', self.release)
+            for component in os.listdir(dist_path):
+                component_path = join(dist_path, component)
+                if not isdir(component_path):
+                    continue
+
+                for binary in os.listdir(component_path):
+                    if binary == 'binary-all':
+                        continue
+                    archs.add(binary.replace('binary-', ''))
+
+            return archs
+
+        components_dir = join(self.path, self.pool)
+        release_dir = join('dists', self.release)
+
+        release = join(self.path, release_dir, 'Release')
+        release_gpg = join(self.path, release_dir, 'Release.gpg')
+
+        for path in (release, release_gpg):
+            if exists(path):
+                os.remove(path)
+
+        hashes = self._archive_cmd('release', release_dir)
+        with open(release, "w") as fob:
+            fob.writelines(
+                [f"Origin: {self.origin}\n",
+                 f"Label: {self.origin}\n",
+                 f"Suite: {self.release}\n",
+                 f"Version: {self.version}\n",
+                 f"Codename: {self.release}\n",
+                 f"Architectures: {' '.join(get_archs())}\n",
+                 f"Components: {' '.join(os.listdir(components_dir))}\n",
+                 f"Description: {self.origin} {self.release} {self.version}\n",
+                 f"{hashes}\n"])
+
+        if gpgkey:
+            subprocess.run(["gpg", "-abs", f"-u {gpgkey}",
+                            f"-o {release_gpg}", release],
+                           text=True)
```

Comment:
I suspect this is incorrect and you intended to write:

```python
            subprocess.run(["gpg", "-abs", "-u", gpgkey,
                            "-o", release_gpg, release],
                           text=True)
```

Reply:
Hmmm, fair call. I have tested it and it works ok, in as much as it runs and appears to do what it's meant to. But this may well just be silently failing...

Reply:
Afaik this would be equivalent to running:

```bash
gpg -abs "-u $gpgkey" "-o $release_gpg", "$release"
```

if that works, then this should work, if that doesn't work, then those changes need to be applied.

Reply:
It turns out that this was failing, but only because it includes the space between the switch and the data. I could work around that, but I'll follow your suggestion as it's then more consistent with existing TurnKey code; plus is easier reading.